core/gesture.js

```diff
@@ -0,0 +1,545 @@
+ * fields have not yet been set.
+ * @return {boolean} true if anything is being dragged.
+ */
+Blockly.Gesture.prototype.updateIsDragging_ = function() {
+  // TODO: Assert that the most recent event was a move?
+  goog.asserts.assert(!this.isDragging(),
+      'Don\'t call updateIsDragging_ when a drag is already in progress.');
+
+  // First check if it was a block drag.
+  if (this.updateIsDraggingBlock_()) {
+    return true;
+  }
+
+  // Then check if it's a workspace drag.
+  var wsMovable = this.flyout_ ?
+      this.flyout_ && (this.flyout_.scrollbar_ != null) &&
+      this.flyout_.scrollbar_.isVisible() :
+      this.startWorkspace_ && this.startWorkspace_.isDraggable();
+  if (wsMovable && this.hasExceededDragRadius_) {
+    this.isDraggingWorkspace_ = true;
+
+    if (this.flyout_) {
+      this.workspaceDragger_ = new Blockly.FlyoutDragger(this.flyout_);
+    } else {
+      this.workspaceDragger_ = new Blockly.WorkspaceDragger(this.startWorkspace_);
+    }
+
+    this.workspaceDragger_.startDrag();
+    return true;
+  }
+  return false;
+};
+
+/**
+ * Start a gesture: update the workspace to indicate that a gesture is in
+ * progress and bind mousemove and mouseup handlers.
+ * @param {!Event} e A mouse down or touch start event.
+ */
+Blockly.Gesture.prototype.doStart = function(e) {
+  // TODO: Blockly.longStart_()
+  this.startWorkspace_.updateScreenCalculationsIfScrolled();
+  this.startWorkspace_.markFocused();
+  this.mostRecentEvent_ = e;
+
+  Blockly.hideChaff(!!this.flyout_);
+
+  if (Blockly.utils.isRightButton(e)) {
+    this.handleRightClick(e);
+    return;
+  }
+  if (this.startBlock_) {
+    this.startBlock_.select();
+  }
+
+  this.mouseDownXY_ = new goog.math.Coordinate(e.clientX, e.clientY);
+
+  this.onMoveWrapper_ = Blockly.bindEvent_(
+      document, 'mousemove', null, this.handleMove.bind(this));
+
+  this.onUpWrapper_ = Blockly.bindEvent_(
+      document, 'mouseup', null, this.handleUp.bind(this));
+
+  e.preventDefault();
+  e.stopPropagation();
+};
+
+/**
+ * Handle a mouse move or touch move event.
+ * @param {!Event} e A mouse move or touch move event.
+ */
+Blockly.Gesture.prototype.handleMove = function(e) {
+  this.update_(e);
+  // TODO: I should probably only call this once, when first exceeding the drag
+  // radius.
+  if (this.hasExceededDragRadius_) {
+    Blockly.longStop_();
+  }
+  if (this.isDraggingWorkspace_) {
+    // Move the visible workspace
+    this.workspaceDragger_.drag(this.currentDragDeltaXY_);
+  } else if (this.isDraggingBlock_) {
+    // Move the dragging block.
+    this.blockDragger_.dragBlock(this.mostRecentEvent_,
+        this.currentDragDeltaXY_);
+  }
+  e.preventDefault();
+  e.stopPropagation();
+};
+
+/**
+ * Handle a mouse up or touch end event.
+ * @param {!Event} e A mouse up or touch end event.
+ */
+Blockly.Gesture.prototype.handleUp = function(e) {
+  this.update_(e);
+  Blockly.longStop_();
+  if (this.isDraggingBlock_) {
+    // Terminate block drag.
```

Comment:
these comments don't seem to add much given that the methods are essentially named the same.  A comment here about why this ordering is important might be useful though.